Add controllers

87stels87 opened the pull request.

87stels87 committed:
13 спринт. попытка 1.

87stels87 committed:
13 спринт. попытка 2. (код-стайл)

87stels87 committed:
13 спринт. попытка 3. (код-стайл)

Sla-als commented on src/main/java/ru/practicum/shareit/item/ItemController.java:
```diff
@@ -1,12 +1,76 @@
+    @ExceptionHandler
+    @ResponseStatus(HttpStatus.NOT_FOUND)
+    public Map<String, String> handle(final NotFoundException e) {
+        return Map.of(
```
Желательно добавить больше информации в логи, особенно в случае ошибок. Это поможет при отладке. Например, можно логировать исключения прямо в обработчиках исключений.

Sla-als replied:
```
@ExceptionHandler
@ResponseStatus(HttpStatus.NOT_FOUND)
public Map<String, String> handle(final NotFoundException e) {
    log.error("Обработка исключения NotFoundException", e);
    return Map.of("Сообщение об ошибке", e.getMessage());
}

```

Sla-als replied:
При возвращении сообщения об ошибке, ключ "Сообщение об ошибке" лучше оставить на английском ("error_message"), поскольку это общепринятый стандарт

Sla-als commented on src/main/java/ru/practicum/shareit/item/ItemService.java:
```diff
@@ -0,0 +1,56 @@
+        return ItemMapper.toItemDto(itemStorage.getItemInfo(itemId));
+    }
+
+    public ItemDto addNewItem(ItemDto itemDto, Long userId) {
```
В методе addNewItem дважды производится проверка itemDto.getName() == null и itemDto.getDescription() == null, а также itemDto.getName().isBlank() и itemDto.getDescription().isBlank(). Это может быть сокращено до одной проверки с использованием оператора OR

Sla-als commented on src/main/java/ru/practicum/shareit/item/ItemService.java:
```diff
@@ -0,0 +1,56 @@
+        return ItemMapper.toItemDto(itemStorage.add(ItemMapper.toItem(itemDto, user.get(), null)));
+    }
+
+    public ItemDto changeItem(Long itemId, Long userId, ItemDto itemDto) {
```
В методе changeItem дважды вызывается userStorage.getById(userId). Это может быть оптимизировано, сохраняя результат первого вызова в переменной

Sla-als commented on src/main/java/ru/practicum/shareit/item/storage/InMemoryItemStorage.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public Collection<Item> getById(Long userId) {
```
Метод getById с названием предполагает, что он возвращает элементы по идентификатору, но на самом деле он возвращает все элементы. Это может вызвать путаницу. Рекомендуется переименовать этот метод, например, в getAll.

Sla-als commented on src/main/java/ru/practicum/shareit/item/storage/InMemoryItemStorage.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public Collection<Item> getByKeyWords(String text) {
```
Проверка if (text.isBlank()) return List.of(); в методе getByKeyWords может быть ненужной, поскольку при поиске с пустым текстом мы все равно получим пустой список после применения фильтров

Sla-als commented on src/main/java/ru/practicum/shareit/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Override
+    public User update(User user, Long userId) {
```
В методе update, вам не нужно вызывать user.setId(userId); после того, как вы уже получили пользователя по этому идентификатору. Вы уже установили идентификатор пользователя и его не нужно повторно устанавливать.

Sla-als replied:
возможно, стоит сначала проверить, есть ли уже пользователь с таким же адресом электронной почты, а затем обновлять данные пользователя. Таким образом, можно избежать ненужных обновлений.

Sla-als commented on src/main/java/ru/practicum/shareit/user/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    @Override
+    public Optional<User> getById(Long id) {
```
В методе getById, вам не обязательно проверять, является ли пользователь null. Вы можете напрямую возвращать Optional.ofNullable(users.get(id));

Sla-als commented on src/main/java/ru/practicum/shareit/user/UserService.java:
```diff
@@ -0,0 +1,42 @@
+
+    public UserDto getUserById(Long userId) {
+        return UserMapper.toUserDto(inMemoryUserStorage.getById(userId)
+                .orElseThrow(() -> new ValidationException("Такого id нет")));
```
Ваш код может быть более однородным, если вы будете использовать аналогичные исключения в аналогичных случаях. Например, в классе ItemService вы выбрасываете NotFoundException при попытке найти пользователя по идентификатору, который не существует. В классе UserService вы выбрасываете ValidationException в аналогичной ситуации. Стоит выбрать один подход

Sla-als reviewed (commented):
Здравствуйте! Ваш код в целом выглядит неплохо, но есть несколько замечаний и предложений по улучшению

87stels87 committed:
13 спринт. попытка 4.
Изменил в контроллере ExceptionHandler. Добавил логирование и вывод сообщения как error_message

87stels87 committed:
13 спринт. попытка 4.
Переименовал метод getById в getAllItemsByUserId

87stels87 committed:
13 спринт. попытка 4.
Заменил проверку на null конструкцией Optional.ofNullable(users.get(id)) в методе Optional<User> getById(Long id)

87stels87 committed:
13 спринт. попытка 4.
В классе UserService заменил на new NotFoundException("Такого id нет")));

87stels87 committed:
13 спринт. попытка 4.
Вынес в переменную optionalUsersById повторяющийся код в методе changeItem

87stels87 committed:
13 спринт. попытка 4.
Убрал дублирование в методе addNewItem(ItemDto itemDto, Long userId)

87stels87 committed:
13 спринт. попытка 4.
Поправил код стайл

Sla-als commented:
Отличная работа, все замечания, полученные на первом этапе учтены и исправлены, зачёт 🎉
